File: BlockPartyWindowsStore/ScreenManagement/ScreenManager.cs
using BlockPartyWindowsStore.Screens;
using BlockPartyWindowsStore.Utilities;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.UI.Core;
using Windows.UI.ViewManagement;

namespace BlockPartyWindowsStore
{
    class ScreenManager
    {
        public Game Game;
        
        List<Screen> screens = new List<Screen>();
        List<Screen> screensToUpdate = new List<Screen>();
        Screen screenToLoad;
        UnsupportedViewStateScreen unsupportedViewStateScreen;
        bool loadingScreen = false;        

        public ScreenManager(Game game)
        {
            Game = game;

            // Setup an unsupported view state screen to show when the game is snapped
            unsupportedViewStateScreen = new UnsupportedViewStateScreen(this);
        }

        public void LoadContent()
        {
            unsupportedViewStateScreen.LoadContent();
        }

        public void AddScreen(Screen screen)
        {
            screen.LoadContent();

            screens.Add(screen);
        }

        public void RemoveScreen(Screen screen)
        {
            screen.UnloadContent();

            screens.Remove(screen);
        }

        public void LoadScreen(Screen screen)
        {
            // Transition out of the current screens
            foreach (Screen s in screens)
            {
                s.TransitionOff();
            }

            screenToLoad = screen;
            loadingScreen = true;
        }

        public void Update(GameTime gameTime)
        {
            // Only update the screen if the app isn't snapped
            if (Game.ApplicationViewState != ApplicationViewState.Snapped)
            {
                // Read input
                Game.InputManager.Update();

                // Make a copy of the list of screens to allow for adds/removes
                // not interfering with updates
                screensToUpdate.Clear();

                foreach (Screen screen in screens)
                {
                    screensToUpdate.Add(screen);
                }

                while (screensToUpdate.Count > 0)
                {
                    // Pop the topmost screen off the stack
                    Screen screen = screensToUpdate[screensToUpdate.Count - 1];
                    screensToUpdate.RemoveAt(screensToUpdate.Count - 1);

                    // Update the screen
                    screen.Update(gameTime);
                    screen.HandleInput(gameTime);
                }

                if (loadingScreen)
                {
                    if (screens.Count == 0)
                    {
                        AddScreen(screenToLoad);
                        loadingScreen = false;
                    }
                }
            }
            else
            {
                unsupportedViewStateScreen.Update(gameTime);
            }
        }

        public void Draw(GameTime gameTime)
        {
            Game.GraphicsDevice.Clear(ClearOptions.Stencil | ClearOptions.Target, Color.Transparent, 0, 0);

            // Only draw the screen if the app isn't snapped
            if (Game.ApplicationViewState != ApplicationViewState.Snapped)
            {
                // Draw screens
                foreach (Screen screen in screens)
                {
                    screen.Draw(gameTime);
                }
            }
            else
            {
                unsupportedViewStateScreen.Draw(gameTime);
            }
        }
    }
}

File: BlockPartyWebRole/Client/ScreenManager.ts
/// <reference path="GameScreen.ts" />
/// <reference path="../Scripts/typings/jquery/jquery.d.ts" />
/// <reference path="GraphicsManager.ts" />
/// <reference path="Viewport.ts" />
/// <reference path="Game.ts" />

class ScreenManager {
    public Game: Game;

    private screens: Array<GameScreen> = new Array<GameScreen>();
    private screensToUpdate: Array<GameScreen> = new Array<GameScreen>();
    private screenToLoad: GameScreen;
    private loadingScreen: boolean;

    constructor(game: Game) {
        this.Game = game;    
    }

    public AddScreen(screen: GameScreen) {
        screen.LoadContent();

        this.screens.push(screen);
    }

    public RemoveScreen(screen: GameScreen) {
        screen.UnloadContent();

        var index = this.screens.indexOf(screen);
        this.screens.splice(index);
    }

    public LoadScreen(screen: GameScreen) {
        // Transition out of the current screens
        this.screens.forEach(function (screen: GameScreen, index: number, screens: Array<GameScreen>) {
            screen.TransitionOff();
        });

        this.screenToLoad = screen;
        this.loadingScreen = true;
    }

    public Update(gameTime: GameTime) {
        // Read input
        this.Game.InputManager.Update();

        // Make a copy of the list of screens to allow for adds/removes not interfering w/ updates
        this.screensToUpdate.splice(0, this.screensToUpdate.length);

        this.screens.forEach(function (screen: GameScreen, index: number, screens: Array<GameScreen>) {
            this.screensToUpdate.push(screen);
        }, this);

        while (this.screensToUpdate.length > 0) {
            // Pop the topmost screen off the stack
            var screen: GameScreen = this.screensToUpdate.pop();

            // Update the screen
            screen.Update(gameTime);
            screen.HandleInput(gameTime);
        }

        if (this.loadingScreen) {
            if (this.screens.length == 0) {
                this.AddScreen(this.screenToLoad);
                this.loadingScreen = false;
            }
        }
    }

    public Draw(gameTime: GameTime) {
        this.Game.GraphicsManager.Clear();

        // Draw screens
        this.screens.forEach(function (screen: GameScreen, index: number, screens: Array<GameScreen>) {
            screen.Draw(gameTime);
        });
    }
}